File: android/app/src/main/java/com/aria/cookie/aria/Patterns.kt
package com.aria.cookie.aria

import com.aria.cookie.core.TimelineEvent
import kotlinx.serialization.Serializable
import kotlin.math.ln
import kotlin.math.sqrt

/**
 * ARIA M3 · Patrones temporales de tu vida (puerto fiel de internal/temporal).
 *
 * "Cuando pasa A, en las siguientes W horas suele pasar B", comparado con la
 * misma hora del día en los días sin A (control por horario), con cota de
 * Wilson. Mismos resultados que Go sobre testdata/aria/timeline.json.
 */
@Serializable
data class Pattern(
    val cause: String,
    val effect: String,
    val windowHours: Int,
    val count: Int,
    val support: Int,
    val probability: Double,
    val baseline: Double,
    val lift: Double,
    val medianLagHours: Double,
    val score: Double,
)

object Patterns {
    private const val HOUR = 3_600_000L
    private const val DAY = 24 * HOUR
    private const val MIN_CONTROLS = 3

    data class Config(
        val windows: List<Int> = listOf(3, 24),
        val minCount: Int = 3,
        val minSupport: Int = 3,
        val minProb: Double = 0.5,
        val minLift: Double = 2.0,
        val maxSymbols: Int = 60,
        val max: Int = 12,
    )

    private val reSleep = Regex("\\((\\d+)h")
    private val reAppMin = Regex("\\s+\\d+ min$")
    private val reMsg = Regex("^mensaje de (.+?)(\\s*\\(.*\\))?$")
    private val rePhoto = Regex("^(\\d+) fotos")

    fun symbol(e: TimelineEvent): String {
        val t = e.text.trim().lowercase()
        return when (e.kind) {
            "lugar" -> "lugar:" + t.removePrefix("llegaste a ")
            "actividad" -> "actividad:$t"
            "movimiento" -> "movimiento:$t"
            "calendario" -> "evento:$t"
            "música" -> "música:$t"
            "estrés" -> "estrés:$t"
            "sueño" -> reSleep.find(t)?.let { if (it.groupValues[1].toInt() < 6) "sueño:corto" else "sueño:normal" } ?: ""
            "app" -> "app:" + reAppMin.replace(t, "")
            "mensaje" -> reMsg.find(t)?.let { "mensaje:" + it.groupValues[1].trim() } ?: ""
            "fotos" -> rePhoto.find(t)?.let { if (it.groupValues[1].toInt() >= 8) "fotos:muchas" else "" } ?: ""
            else -> ""
        }
    }

    fun mine(events: List<TimelineEvent>, cfg: Config = Config()): List<Pattern> {
        val occ = mutableMapOf<String, MutableList<Long>>()
        var first = 0L
        var last = 0L
        for (e in events) {
            val s = symbol(e)
            if (s.isEmpty()) continue
            occ.getOrPut(s) { mutableListOf() } += e.at
            if (first == 0L || e.at < first) first = e.at
            if (e.at > last) last = e.at
        }
        if (occ.size < 2 || last - first < 2 * DAY) return emptyList()
        val clean = occ.mapValues { (_, ts) -> dedupe(ts.sorted(), 30 * 60_000L) }
        val symbols = clean.keys.sortedWith(compareByDescending<String> { clean[it]!!.size }.thenBy { it }).take(cfg.maxSymbols)

        val best = linkedMapOf<Pair<String, String>, Pattern>()
        for (w in cfg.windows) {
            val wMs = w * HOUR
            for (a in symbols) {
                val occA = clean[a]!!
                if (occA.size < cfg.minCount) continue
                val controls = controlTimes(occA, first, last)
                if (controls.size < MIN_CONTROLS) continue
                for (b in symbols) {
                    if (a == b) continue
                    val (support, lags) = followed(occA, clean[b]!!, wMs)
                    if (support < cfg.minSupport) continue
                    val n = occA.size
                    val p = support.toDouble() / n
                    val ctrlHits = followed(controls, clean[b]!!, wMs).first
                    val base = (ctrlHits + 0.5) / (controls.size + 1)
                    val lift = p / base
                    if (p < cfg.minProb || lift < cfg.minLift) continue
                    val score = wilsonLower(support, n) * (ln(lift) / ln(2.0))
                    val key = a to b
                    val old = best[key]
                    if (old != null && old.score >= score) continue
                    best[key] = Pattern(a, b, w, n, support, p, base, lift, median(lags) / HOUR, score)
                }
            }
        }
        return best.values.sortedWith(compareByDescending<Pattern> { it.score }.thenBy { it.cause + it.effect }).take(cfg.max)
    }

    private fun followed(xs: List<Long>, ys: List<Long>, w: Long): Pair<Int, List<Double>> {
        var support = 0
        val lags = mutableListOf<Double>()
        for (a in xs) {
            val i = upperBound(ys, a)
            if (i < ys.size && ys[i] - a <= w) { support++; lags += (ys[i] - a).toDouble() }
        }
        return support to lags
    }

    private fun controlTimes(xs: List<Long>, first: Long, last: Long): List<Long> {
        val out = mutableListOf<Long>()
        for (a in xs) {
            var t = a - ((a - first) / DAY) * DAY
            while (t <= last) {
                if (t != a && !near(xs, t, 2 * HOUR)) out += t
                t += DAY
            }
        }
        return out.sorted()
    }

    private fun near(ts: List<Long>, t: Long, d: Long): Boolean {
        val i = lowerBound(ts, t - d)
        return i < ts.size && ts[i] <= t + d
    }

    /** Primer índice con valor > x. */
    private fun upperBound(xs: List<Long>, x: Long): Int {
        var lo = 0; var hi = xs.size
        while (lo < hi) { val m = (lo + hi) ushr 1; if (xs[m] > x) hi = m else lo = m + 1 }
        return lo
    }

    /** Primer índice con valor ≥ x. */
    private fun lowerBound(xs: List<Long>, x: Long): Int {
        var lo = 0; var hi = xs.size
        while (lo < hi) { val m = (lo + hi) ushr 1; if (xs[m] >= x) hi = m else lo = m + 1 }
        return lo
    }

    private fun dedupe(ts: List<Long>, gap: Long): List<Long> {
        val out = mutableListOf<Long>()
        for (t in ts) if (out.isEmpty() || t - out.last() >= gap) out += t
        return out
    }

    private fun median(xs: List<Double>): Double = if (xs.isEmpty()) 0.0 else xs.sorted()[xs.size / 2]

    private fun wilsonLower(k: Int, n: Int): Double {
        if (n == 0) return 0.0
        val z = 1.645
        val p = k.toDouble() / n
        val nn = n.toDouble()
        val den = 1 + z * z / nn
        val center = p + z * z / (2 * nn)
        val margin = z * sqrt(p * (1 - p) / nn + z * z / (4 * nn * nn))
        return (center - margin) / den
    }

    fun describe(p: Pattern): String {
        val whenTxt = if (p.windowHours >= 24) "en el día siguiente" else "en las horas siguientes"
        val strength = if (p.lift > 10) "mucho más de lo normal" else "%.1f× más de lo normal".format(java.util.Locale.US, p.lift)
        return "Cuando ${human(p.cause)}, ${human(p.effect)} $whenTxt (${p.support} de ${p.count} veces; $strength)"
    }

    fun human(sym: String): String {
        val kind = sym.substringBefore(':')
        val v = sym.substringAfter(':')
        return when (kind) {
            "lugar" -> "vas a $v"
            "actividad" -> "toca $v"
            "movimiento" -> mapOf("caminar" to "caminas", "correr" to "corres", "bici" to "vas en bici", "en coche" to "vas en coche")[v] ?: v
            "evento" -> "tienes «$v»"
            "música" -> "escuchas a $v"
            "estrés" -> "tu estrés está $v"
            "sueño" -> if (v == "corto") "duermes poco" else "duermes bien"
            "app" -> "usas $v"
            "mensaje" -> "hablas con $v"
            "fotos" -> "haces muchas fotos"
            else -> sym
        }
    }
}

File: internal/temporal/patterns.go
// Package temporal es la mitad temporal del Módulo 3 de ARIA: descubre
// patrones reales en tu vida a partir de la línea de tiempo de Cookie.
//
// Busca relaciones "cuando pasa A, en las siguientes W horas suele pasar B"
// y solo las acepta si B ocurre después de A bastante más que a esa misma
// hora en los días sin A (control por horario), con suficientes casos y una
// cota inferior de Wilson para no fiarse de rachas. Es correlación temporal, no prueba de
// causa: por eso las llama "patrones" y muestra cuántas veces se cumplió.
package temporal

import (
	"fmt"
	"math"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

// Event algo que Cookie anotó en tu línea de tiempo.
type Event struct {
	At   int64  `json:"at"` // milisegundos Unix
	Kind string `json:"kind"`
	Text string `json:"text"`
}

// Pattern relación temporal descubierta.
type Pattern struct {
	Cause       string  `json:"cause"`
	Effect      string  `json:"effect"`
	WindowHours int     `json:"window_hours"`
	Count       int     `json:"count"`   // veces que ocurrió la causa
	Support     int     `json:"support"` // veces que después vino el efecto
	Probability float64 `json:"probability"`
	Baseline    float64 `json:"baseline"`
	Lift        float64 `json:"lift"`
	MedianLagH  float64 `json:"median_lag_hours"`
	Score       float64 `json:"score"`
}

// Config parámetros del minado.
type Config struct {
	Windows    []int // horas
	MinCount   int
	MinSupport int
	MinProb    float64
	MinLift    float64
	MaxSymbols int
	Max        int
}

// DefaultConfig valores prudentes para datos personales.
func DefaultConfig() Config {
	return Config{Windows: []int{3, 24}, MinCount: 3, MinSupport: 3, MinProb: 0.5, MinLift: 2, MaxSymbols: 60, Max: 12}
}

var (
	reSleep = regexp.MustCompile(`\((\d+)h`)
	reAppMn = regexp.MustCompile(`\s+\d+ min$`)
	reMsg   = regexp.MustCompile(`^mensaje de (.+?)(\s*\(.*\))?$`)
	rePhoto = regexp.MustCompile(`^(\d+) fotos`)
)

// Symbol convierte un evento en un símbolo comparable ("lugar:gimnasio").
// Devuelve "" si el evento no sirve para buscar patrones.
func Symbol(e Event) string {
	t := strings.ToLower(strings.TrimSpace(e.Text))
	switch e.Kind {
	case "lugar":
		return "lugar:" + strings.TrimPrefix(t, "llegaste a ")
	case "actividad":
		return "actividad:" + t
	case "movimiento":
		return "movimiento:" + t
	case "calendario":
		return "evento:" + t
	case "música":
		return "música:" + t
	case "estrés":
		return "estrés:" + t
	case "sueño":
		if m := reSleep.FindStringSubmatch(t); m != nil {
			if h, _ := strconv.Atoi(m[1]); h < 6 {
				return "sueño:corto"
			}
			return "sueño:normal"
		}
	case "app":
		return "app:" + reAppMn.ReplaceAllString(t, "")
	case "mensaje":
		if m := reMsg.FindStringSubmatch(t); m != nil {
			return "mensaje:" + strings.TrimSpace(m[1])
		}
	case "fotos":
		if m := rePhoto.FindStringSubmatch(t); m != nil {
			if n, _ := strconv.Atoi(m[1]); n >= 8 {
				return "fotos:muchas"
			}
		}
	}
	return ""
}

// Mine descubre patrones en una línea de tiempo.
func Mine(events []Event, cfg Config) []Pattern {
	occ := map[string][]int64{}
	var first, last int64
	for _, e := range events {
		s := Symbol(e)
		if s == "" {
			continue
		}
		occ[s] = append(occ[s], e.At)
		if first == 0 || e.At < first {
			first = e.At
		}
		if e.At > last {
			last = e.At
		}
	}
	if len(occ) < 2 || last-first < 2*24*hourMs {
		return nil
	}
	symbols := make([]string, 0, len(occ))
	for s, ts := range occ {
		sort.Slice(ts, func(i, j int) bool { return ts[i] < ts[j] })
		occ[s] = dedupe(ts, 30*60_000) // la misma cosa repetida en media hora cuenta una vez
		symbols = append(symbols, s)
	}
	sort.Slice(symbols, func(i, j int) bool {
		if len(occ[symbols[i]]) != len(occ[symbols[j]]) {
			return len(occ[symbols[i]]) > len(occ[symbols[j]])
		}
		return symbols[i] < symbols[j]
	})
	if len(symbols) > cfg.MaxSymbols {
		symbols = symbols[:cfg.MaxSymbols]
	}

	best := map[[2]string]Pattern{}
	for _, w := range cfg.Windows {
		wMs := int64(w) * hourMs
		for _, a := range symbols {
			if len(occ[a]) < cfg.MinCount {
				continue
			}
			controls := controlTimes(occ[a], first, last)
			if len(controls) < minControls {
				continue // pasa (casi) todos los días: no hay con qué comparar
			}
			for _, b := range symbols {
				if a == b {
					continue
				}
				support, lags := followed(occ[a], occ[b], wMs)
				if support < cfg.MinSupport {
					continue
				}
				n := len(occ[a])
				p := float64(support) / float64(n)
				ctrlHits, _ := followed(controls, occ[b], wMs)
				// Tasa base: misma hora del día, en días sin la causa (suavizada).
				base := (float64(ctrlHits) + 0.5) / (float64(len(controls)) + 1)
				lift := p / base
				if p < cfg.MinProb || lift < cfg.MinLift {
					continue
				}
				score := wilsonLower(support, n) * math.Log2(lift)
				key := [2]string{a, b}
				if old, ok := best[key]; ok && old.Score >= score {
					continue
				}
				best[key] = Pattern{
					Cause: a, Effect: b, WindowHours: w, Count: n, Support: support,
					Probability: p, Baseline: base, Lift: lift, MedianLagH: median(lags) / float64(hourMs), Score: score,
				}
			}
		}
	}
	out := make([]Pattern, 0, len(best))
	for _, p := range best {
		out = append(out, p)
	}
	sort.Slice(out, func(i, j int) bool {
		if out[i].Score != out[j].Score {
			return out[i].Score > out[j].Score
		}
		return out[i].Cause+out[i].Effect < out[j].Cause+out[j].Effect
	})
	if len(out) > cfg.Max {
		out = out[:cfg.Max]
	}
	return out
}

const hourMs = int64(3_600_000)

// followed cuántas veces A fue seguido por B en (0, w], y el retraso de cada vez.
func followed(as, bs []int64, w int64) (int, []float64) {
	support := 0
	var lags []float64
	for _, a := range as {
		i := sort.Search(len(bs), func(i int) bool { return bs[i] > a })
		if i < len(bs) && bs[i]-a <= w {
			support++
			lags = append(lags, float64(bs[i]-a))
		}
	}
	return support, lags
}

// minControls días de control mínimos para poder comparar.
const minControls = 3

// controlTimes momentos de control: la misma hora del día que cada causa, en
// otros días del periodo en que la causa NO ocurrió cerca de esa hora. Así la
// comparación descuenta tu horario (lo que pasa a esa hora de todos modos).
func controlTimes(as []int64, first, last int64) []int64 {
	const day = 24 * hourMs
	var out []int64
	for _, a := range as {
		for t := a - ((a-first)/day)*day; t <= last; t += day {
			if t == a || near(as, t, 2*hourMs) {
				continue
			}
			out = append(out, t)
		}
	}
	sort.Slice(out, func(i, j int) bool { return out[i] < out[j] })
	return out
}

func near(ts []int64, t, d int64) bool {
	i := sort.Search(len(ts), func(i int) bool { return ts[i] >= t-d })
	return i < len(ts) && ts[i] <= t+d
}

func dedupe(ts []int64, gap int64) []int64 {
	out := ts[:0:0]
	for _, t := range ts {
		if len(out) == 0 || t-out[len(out)-1] >= gap {
			out = append(out, t)
		}
	}
	return out
}

func median(xs []float64) float64 {
	if len(xs) == 0 {
		return 0
	}
	s := append([]float64(nil), xs...)
	sort.Float64s(s)
	return s[len(s)/2]
}

// wilsonLower cota inferior (90%) de la proporción k/n.
func wilsonLower(k, n int) float64 {
	if n == 0 {
		return 0
	}
	z := 1.645
	p := float64(k) / float64(n)
	nn := float64(n)
	den := 1 + z*z/nn
	center := p + z*z/(2*nn)
	margin := z * math.Sqrt(p*(1-p)/nn+z*z/(4*nn*nn))
	return (center - margin) / den
}

// Describe frase en español para mostrar el patrón.
func Describe(p Pattern) string {
	when := "en las horas siguientes"
	if p.WindowHours >= 24 {
		when = "en el día siguiente"
	}
	strength := fmt.Sprintf("%.1f× más de lo normal", p.Lift)
	if p.Lift > 10 {
		strength = "mucho más de lo normal"
	}
	return fmt.Sprintf("Cuando %s, %s %s (%d de %d veces; %s)",
		Human(p.Cause), Human(p.Effect), when, p.Support, p.Count, strength)
}

// Human traduce un símbolo a lenguaje natural.
func Human(sym string) string {
	kind, val, _ := strings.Cut(sym, ":")
	switch kind {
	case "lugar":
		return "vas a " + val
	case "actividad":
		return "toca " + val
	case "movimiento":
		return map[string]string{"caminar": "caminas", "correr": "corres", "bici": "vas en bici", "en coche": "vas en coche"}[val]
	case "evento":
		return "tienes «" + val + "»"
	case "música":
		return "escuchas a " + val
	case "estrés":
		return "tu estrés está " + val
	case "sueño":
		if val == "corto" {
			return "duermes poco"
		}
		return "duermes bien"
	case "app":
		return "usas " + val
	case "mensaje":
		return "hablas con " + val
	case "fotos":
		return "haces muchas fotos"
	}
	return sym
}
